### PR TITLE
Fixes Issue #206 duplicated argument name

### PR DESCRIPTION
**Pull request by wflanagan**
as well as minor changes for compatibility with non-Rails environments, including: 

1) Supports CEQUEL_MODELS_PATH env var so you can point to a destination besides app/models for loading of files.  Does not handle namespaced models in this folder yet, but its easy to do from here. I might do that in a later request.  
2) Downgrade to Rails 3.2 compatibility to be in compliance with documentation, which states that it supports > 3.1.
3) Adds a check to ensure that failed rspec tests on one run don't mess up the runs of subsequent tests.  

How the fix is implemented:
Changes to the duplicated argument name were fixed by changing accessors from Arrays to Hashes, and then using the hash.values to set the arrays.  I also implemented minor checks to ensure that some weird edge case won't create further problems. 
I also broke out some fatter methods in the table_writer.rb to make it easier to make changes in the future.   

Notes: 
It was several days in before I realized that the CEQUEL_TEST_HOST env var was available. This should be made more prominent in the documentation, as its required for Docker compatibility.

### CODE REVIEW

File: lib/cequel/schema/table.rb [hunk 1]
```diff
@@ -157,11 +157,26 @@ def data_columns
...
+          index_name = options[:index]
+          index_name = :"#{@name}_#{name}_idx" if index_name == true
+          DataColumn.new(name, type(type), index_name)
+            .tap do |column|
+              data_columns_hash[name] = add_column(column)
+            end
+        else
+          # return the data column like we would if we built it
+          data_columns_hash[name]
+        end
```
Comment (pezra) [inline lib/cequel/schema/table.rb]: How would you feel about raising an exception if the column being defined didn't match the previously defined column? That might prevent some subtle errors.

File: lib/cequel/record/tasks.rb [hunk 2]
```diff
@@ -63,12 +63,15 @@ def drop!
     puts "Dropped keyspace #{Cequel::Record.connection.name}"
   end
 
+  def models_dir_path
+    ENV['CEQUEL_MODELS_PATH'] || "#{File.expand_path('app/models', project_root)}/"
```
Comment (pezra) [inline lib/cequel/record/tasks.rb]: `ENV.fetch('CEQUEL_MODELS_PATH', project_root + "app/models")`? We'd have to ensure that `project_root` was a `Pathname` but that would be a Good Thing.

File: lib/cequel/schema/table.rb [hunk 3]
```diff
@@ -1,5 +1,6 @@
 # -*- encoding : utf-8 -*-
 require 'stringio'
+require 'set'
```
Comment (pezra) [inline lib/cequel/schema/table.rb]: Is the `Set` class used?

File: lib/cequel/schema/table.rb [hunk 4]
```diff
@@ -59,17 +48,50 @@ def initialize(name)
       # @see #add_partition_key
       #
       def add_key(name, type, clustering_order = nil)
-        if @partition_key_columns.empty?
-          unless clustering_order.nil?
-            fail ArgumentError,
-                 "Can't set clustering order for partition key #{name}"
+        if valid_key?(name)
+          if partition_key_columns.empty?
```
Comment (pezra) [inline lib/cequel/schema/table.rb]: 👍 for using reader methods!

**Comment by pezra** [comment]
On the whole the changes in this PR look really good.

There is some sort of DNS issue with rubygems.org so i cannot restart that one failed build at the moment but i will tomorrow. I am pretty sure the it will pass (our tests are a little flakey).

File: lib/cequel/record/associations.rb [hunk 5]
```diff
@@ -100,26 +100,30 @@ module ClassMethods
...
-          if table_schema.key_columns.any?
-            fail InvalidRecordConfiguration,
-                 "belongs_to association must be declared before declaring " \
-                 "key(s)"
-          end
+            # skip adding the belongs to if already defined on a reload
+            unless parent_association.association_class_name.downcase.to_sym == name.to_sym
+              fail InvalidRecordConfiguration,
+                   "Can't declare more than one belongs_to association"
+            end
```
Comment (wflanagan) [inline lib/cequel/record/associations.rb]: Sure, i can do that. I will make the change in the next few days and push it.

File: lib/cequel/schema/table.rb (hunk 3 above)
Comment (wflanagan) [inline lib/cequel/schema/table.rb]: On this.. i saw it was used in the class.. in troubleshooting, i wanted to be sure that things were OK.. i will remove in the revision.

File: lib/cequel/record/tasks.rb (hunk 2 above)
Comment (wflanagan) [inline lib/cequel/record/tasks.rb]: Not sure I get what you're saying on this one. This is operational for me using the CEQUEL_MODELS_PATH to point to an alternative location for the model files.  Help me understand and I'm happy to patch.

File: lib/cequel/schema/table.rb (hunk 1 above)
Comment (wflanagan) [inline lib/cequel/schema/table.rb]: Yeah, I thought about that a lot, and how that would work in the case of migrations. But, then, I decided that made my head hurt.. so I did it this way. I think though that it's a good thought, and I definitely thought it as well. I'll see what I can do on my next pull request.

File: lib/cequel/record/tasks.rb (hunk 2 above)
Comment (pezra) [inline lib/cequel/record/tasks.rb]: I'm saying that i like `Hash#fetch` and `Pathname#+` better than `Hash#[]` and `File.expand_path`. 

It is a small thing, on which reasonable people can disagree, and is not necessary to merge this PR.

**Comment by pezra** [comment]
Any chance i could get this change on 2.0? I'd love to have it.

**Comment by wflanagan** [comment]
I'll see what I can do.. had to move off of Cassandra for a bit.. as soon
as I get back to it (hopefully a few days), I can push an update.

**Comment by wflanagan** [comment]
Hey, what's the status here? Should I use this and try to upgrade, or did you make big changes on 2.0 that make this irrelevant?

File: lib/cequel/schema/table.rb (hunk 1 above)
Comment (wflanagan) [inline lib/cequel/schema/table.rb]: Ok.. I've upgraded to 2.0.3 and I'm implementing these changes.  This is done.

File: lib/cequel/schema/table.rb (hunk 3 above)
Comment (wflanagan) [inline lib/cequel/schema/table.rb]: Done.

File: lib/cequel/record/tasks.rb (hunk 2 above)
Comment (wflanagan) [inline lib/cequel/record/tasks.rb]: I've done it, I think. :-) I set it in the "tell don't ask" style that checks for Rails::Railtie (much more reliable than Rails, as many people monkey patch the Rails module into existence in non-rails systems). If there, will use that folder. If not, will use the Dir.pwd as was used in the previous version. Then, I check the resulting pathname to see if it exists. If not, I raise a NoModelDirectoryError.

**Comment by pezra** [comment]
@wflanagan, the changes for 2.0 are relatively minor, but still breaking, changes. Thanks for moving forward.

**Comment by pezra** [comment]
Something weird is going on with the travis build. I am looking into it. I'll let you know once i know what is up.

**Comment by wflanagan** [comment]
Let me know what they are, and I’ll try to fix.

WF

**Comment by pezra** [comment]
@wflanagan, can you add a commit to this PR? Maybe add a blank line to the end of the README or something?

**Comment by wflanagan** [comment]
Put a space in the Vagrantfile.  Let's see what happens.

**Comment by pezra** [comment]
@wflanagan, just wanted you to know i haven't forgotten this. Hopefully #356 will get this branch to a green build.

**Comment by wflanagan** [comment]
No worries. I’m trucking on my branch.. when it gets in, if it gets in,
I’ll swing to your branch.

WF

**Comment by pezra** [comment]
@wflanagan, i don't know what is up with this branch exactly, but it fails at a much higher frequency than master, or any other branch. That indicates there is a real problem in the code. It is probably a race condition in the code, or order dependency in the tests given the intermittency of failure.

I do like the change and if we can get the build green i will merge it, but until then i cannot.

**Comment by wflanagan** [comment]
Of course. Can you point me at a bit of the code that’s problematic? If so,
I’ll take a crack at fixing.

**Comment by pezra** [comment]
I don't currently have any clue as to the root of the problem. The failures are intermittent and not always in the same tests but the failures all seem related to changing the schema.

**Comment by wflanagan** [comment]
Hey, on this.. I"ve updated to Cequel 3.0 myself, changed libraries, etc. and effectively made this irrelevant. Closing.  I'll come back and contribute in other ways.. ;-)

**Pull request closed without merging**